Add examples on plugins usage (#34)
aurora-is-near/near-plugins: examples/access-controllable-examples/README.md

# Example of using the Access Control plugin

An access control mechanism that allows you to specify which groups of users can access certain functions.

```rust
use near_plugins::AccessControllable;
use near_plugins::AccessControlRole;
use near_plugins_derive::access_control;
use near_plugins_derive::access_control_any;
use near_sdk::near_bindgen;
use borsh::{BorshSerialize, BorshDeserialize};

> use near_plugins::events::AsEvent;

Reviewer: ```suggestion
```

This shouldn't be required anymore (ref. #21).

Author: fixed https://github.com/aurora-is-near/near-plugins/pull/34/commits/a334a44069b91318440e1a87c74f219a248b9920

use near_sdk::env;

/// All types of access groups
#[derive(AccessControlRole, Clone, Copy)]
pub enum UsersGroups {
    GroupA,
    GroupB,
}

#[near_bindgen]

> #[access_control(role_type="UsersGroups")]

Author: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/a334a44069b91318440e1a87c74f219a248b9920

#[derive(Default, BorshSerialize, BorshDeserialize)]
struct Counter {
    counter: u64,
}

#[near_bindgen]
impl Counter {
    /// In the constructor we set up a super admin, 
    /// which can control the member lists of all user groups
    #[init]
    pub fn new() -> Self {
        let mut contract: Counter = Self{
            counter: 0,
            __acl: __Acl::default(),
        };

>         contract.__acl.init_super_admin(&near_sdk::env::predecessor_account_id());

Author: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/a334a44069b91318440e1a87c74f219a248b9920

        contract
    }

    /// unprotected function, every one can call this function
    pub fn unprotected(&mut self) {
        self.counter += 1;
    }

    /// only the users from GroupA can call this method
    #[access_control_any(roles(UsersGroups::GroupA))]
    pub fn level_a_incr(&mut self) {
        self.counter += 1;
    }

    /// only the users from GroupA or GroupB can call this method
    #[access_control_any(roles(UsersGroups::GroupA, UsersGroups::GroupB))]
    pub fn level_ab_incr(&mut self) {
        self.counter += 1;
    }

    /// view method for get current counter value, every one can use it
    pub fn get_counter(&self) -> u64 {
        self.counter
    }
}
```

## The contract methods description
### acl_storage_prefix
`acl_storage_prefix` is a method that returns the common prefix of keys for storing the members and the admins of groups.

> `__acl` by default.

Author: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/c11247d4f3fc1f49a14fd6560235b8b3b22b4751

```shell
$ near call <CONTRACT_ACCOUNT> acl_storage_prefix --accountId <ALICE_ACCOUNT>
Scheduling a call: <CONTRACT_ACCOUNT>.acl_storage_prefix()
Doing account.functionCall()
Transaction Id <TRANSACTION_ID>
To see the transaction in the transaction explorer, please open this url in your browser
https://explorer.testnet.near.org/transactions/<TRANSACTION_ID>
[ 95, 95, 97, 99, 108 ]
$ python3
>>> print(' '.join(str(b) for b in bytes("__acl", 'utf8')))
95 95 97 99 108
```

> 

Author: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6ce74158b38e3243b03abbe8b25d827601c1547b

> 

Author: @mooori is it possible to remove the super admin somehow?

> 

Reviewer: There is [`revoke_super_admin_unchecked`](https://github.com/aurora-is-near/near-plugins/blob/5579de43c28984d00fa9aebc0a8e755258bba6c2/near-plugins-derive/src/access_controllable.rs#L169-L171), it can be called like this from inside a contract method:

```rust
self.__acl.revoke_super_admin_unchecked(&account_id);
```

### acl_is_super_admin
`acl_is_super_admin` is a _view_ method that checks that account has super admin rights.
Super admin can control the member list of each group and control the admin list for each group.

```shell
$ near view <CONTRACT_ACCOUNT> acl_is_super_admin '{"account_id": "<SUPER_ADMIN_ACCOUNT>" }' 
View call: <CONTRACT_ACCOUNT>.acl_is_super_admin({"account_id": "<SUPER_ADMIN_ACCOUNT>"})
true
```

### acl_add_admin
`acl_add_admin` is a method that adds a new admin for a specific group.
Admins' rights don't allow running group-specific functions, but group admins can control the group member list.

> This method can be run by the admin of a specific group or by the super admin.

Reviewer: ```suggestion
This method can be run by an admin of a specific group or by a super admin.
```

There might be multiple admins and super admins.

Author: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6fd1a4a364b2180df630ee7c4a85b3718c14c89d

```shell
$ near call <CONTRACT_ACCOUNT> acl_add_admin '{"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"}' --accountId <GROUP_A_ADMIN_ACCOUNT>
```

### acl_is_admin

> `acl_is_admin` is a _view_ method that checks if the account has an admin right for the specified group. For super admin, it will return true.

Author: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6fd1a4a364b2180df630ee7c4a85b3718c14c89d

```shell
$ near view <CONTRACT_ACCOUNT> acl_is_admin '{"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"}'
View call: <CONTRACT_ACCOUNT>.acl_is_admin({"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"})
true
```

### acl_revoke_admin
`acl_revoke_admin` is a method that removes the group admin right for a specific account. Can be executed by the admin of this group or by the super admin.

```shell
$ near call <CONTRACT_ACCOUNT> acl_revoke_admin '{"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"}' --accountId <GROUP_A_ADMIN_ACCOUNT>
```

### acl_renounce_admin
`acl_renounce_admin` is a method that removes the group admin right for an account that calls the method.

```shell
$ near call <CONTRACT_ACCOUNT> acl_renounce_admin '{"role": "GroupA"}' --accountId <ALICE_ACCOUNT>
```

After calling that method, Alice will not have the admin right for GroupA anymore.

### acl_revoke_role
`acl_revoke_role` is a method that removes the specified account from the list of the group member.

> Only the group admin or super admin can execute this function.

Author: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6fd1a4a364b2180df630ee7c4a85b3718c14c89d

```shell
$ near call <CONTRACT_ACCOUNT> acl_renounce_role '{"role": "GroupA"}' --accountId <ALICE_ACCOUNT>
```

### acl_grant_role
`acl_grant_role` is a method that adds the account to the group member list. Can be executed only by the group admin or by the super admin.

```shell
$ near call <CONTRACT_ACCOUNT> acl_grant_role '{"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"}' --accountId <GROUP_A_ADMIN_ACCOUNT>
```

### acl_has_role
`acl_has_role` is a _view_ method for checking if the account is a member of the specified group.

```shell
$ near view <CONTRACT_ACCOUNT> acl_has_role '{"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"}'
View call: <CONTRACT_ACCOUNT>.acl_has_role({"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"})
true
```

### acl_has_any_role
`acl_has_any_role` is a _view_ method for checking if an account is a member of at least one of the specified groups.

```shell
$ near view <CONTRACT_ACCOUNT> acl_has_any_role '{"roles": ["GroupA", "GroupB"], "account_id": "<ALICE_ACCOUNT>"}'
View call: <CONTRACT_ACCOUNT>.acl_has_any_role({"roles": ["GroupA", "GroupB"], "account_id": "<ALICE_ACCOUNT>"})
true
```

> 

Author: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6ce74158b38e3243b03abbe8b25d827601c1547b

95 95 97 99 108
```

Example of changing acl storage prefix key:
```rust
#[near_bindgen]
#[access_control(role_type(UsersGroups), storage_prefix = "__custom_prefix")]
#[derive(Default, BorshSerialize, BorshDeserialize)]
struct Counter {
  counter: u64,
}
```

### acl_init_super_admin
`acl_init_super_admin` is a method that adds `account_id` as a super-admin _without_ checking any permissions
in case there are no super-admins. Do nothing if at least one super-admin exists. This function can be used to add a super-admin during contract initialization. 
Moreover, it may provide a recovery mechanism if (mistakenly) all super-admins have been removed. 

**Return value:** the return value indicates whether `account_id` was added as super-admin.

It is `#[private]` in the implementation provided by this trait, i.e. only the contract itself may call this method.

```shell
$ near call <CONTRACT_ACCOUNT> acl_init_super_admin '{"account_id": "<SUPER_ADMIN_ACCOUNT>"}' --accountId <CONTRACT_ACCOUNT>
true
```

If the method succeeds, the following event will be emitted:
```json
{
   "standard":"AccessControllable",
   "version":"1.0.0",
   "event":"super_admin_added",
   "data":{

>       "account":"test.near",

Reviewer: ```suggestion
      "account":"<SUPER_ADMIN_ACCOUNT",
```

Author: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/1a7a34b88d2b4513d1de096b408ef5fc17e46209

For using the method, `level_a_incr` you should be a member of GroupA.  Alice is not a member of any group, so she can't use this method.

```shell
$ near call <CONTRACT_ACCOUNT> level_a_incr --accountId <ALICE_ACCOUNT>
ERROR
$ near view get_counter
0
```

Let's make Alice a member of GroupA.
```shell
$ near call <CONTRACT_ACCOUNT> acl_grant_role '{"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"}' --accountId <CONTRACT_ACCOUNT>
```

Now Alice is a member of GroupA and can call the `level_a_incr` method
```shell
$ near call <CONTRACT_ACCOUNT> level_a_incr --accountId <ALICE_ACCOUNT>
$ near view get_counter
1
```

As well as calls the `level_ab_incr` method, which allowed for both GroupA and GroupB members.
```shell
$ near call <CONTRACT_ACCOUNT> level_ab_incr --accountId <ALICE_ACCOUNT>
$ near view get_counter
2
```

### Admin of the group not a member of the group
Note, the admin of the group may not be a member of this group. For example, the `<CONTRACT_ACCOUNT>` is a super admin, but he
can't execute the `level_a_incr` method. 

```shell
$ near view get_counter
2
$ near call <CONTRACT_ACCOUNT> level_a_incr --accountId <CONTRACT_ACCOUNT>
ERROR
$ near view get_counter
2
```

> ### A few super admins
> There may be two or more super admins. 
> By default, there are no interfaces that allow you to add or remove super admins. 
> You can add the corresponding functions to the contract as follows:

Reviewer: I'd rephrase it a bit, something along those lines:

```suggestion
### Multiple super admins
There may be multiple super admins. 
By default, only `acl_init_super_admin` is exposed on the contract. To add more super admins or to remove super admins, internal functions can be used. 
To make this functionality publicly available, you could add functions like the following to your contract:
```

Author: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/1a7a34b88d2b4513d1de096b408ef5fc17e46209

```rust
/// method for adding new super admin
pub fn add_super_admin(&mut self, new_super_admin_account_id: &AccountId) {

>     ::near_sdk::require!(self.__acl.is_super_admin(&near_sdk::env::predecessor_account_id()),

Reviewer: The trait method is documented, so I'd suggest calling that:

```suggestion
    ::near_sdk::require!(self.acl_is_super_admin(near_sdk::env::predecessor_account_id()),
```

Author: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/1a7a34b88d2b4513d1de096b408ef5fc17e46209

                         "Method can be run only by super admin");
    self.__acl.add_super_admin_unchecked(new_super_admin_account_id); 
}

/// method for removing super admin
pub fn remove_super_admin(&mut self, super_admin_account_id: &AccountId) {

>    ::near_sdk::require!(self.__acl.is_super_admin(&near_sdk::env::predecessor_account_id()),

Reviewer: ```suggestion
   ::near_sdk::require!(self.acl_is_super_admin(near_sdk::env::predecessor_account_id()),
```

Author: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/1a7a34b88d2b4513d1de096b408ef5fc17e46209
